core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp

```diff
@@ -0,0 +1,429 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_PARALLEL_REDUCE_RANGE_HPP
+#define KOKKOS_OPENACC_PARALLEL_REDUCE_RANGE_HPP
+
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <Kokkos_Parallel.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+// Default to catch all non implemented Reducers
+template <class Reducer, class FunctorType, class ExePolicy, class TagType>
+struct OpenACCReductionWrapper {
+  using value_type = typename Reducer::value_type;
+
+#pragma acc routine seq
+  KOKKOS_INLINE_FUNCTION
+  static void init(value_type&) {
+    Kokkos::abort(
+        "[ERROR in reduce()] The given Reducer is not implemented in the "
+        "OpenACC "
+        "backend.\n");
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  static void reduce(value_type&, const ExePolicy&, const FunctorType&) {
+    Kokkos::abort(
+        "[ERROR in reduce()] The given Reducer is not implemented in the "
+        "OpenACC "
+        "backend.\n");
+  }
+};
+
+// Specializations with implemented Reducers
+
+template <class Scalar, class Space, class FunctorType, class TagType,
+          class... Traits>
+struct OpenACCReductionWrapper<Sum<Scalar, Space>, FunctorType,
+                               Kokkos::RangePolicy<Traits...>, TagType> {
+ public:
+  using value_type = typename std::remove_cv<Scalar>::type;
+  using Policy     = Kokkos::RangePolicy<Traits...>;
+
+#pragma acc routine seq
+  KOKKOS_INLINE_FUNCTION
+  static void init(value_type& val) {
+    val = reduction_identity<value_type>::sum();
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  static void reduce(value_type& tmp, const Policy& m_policy,
+                     const FunctorType& m_functor) {
+    const FunctorType a_functor(m_functor);
+    const auto begin = m_policy.begin();
+    const auto end   = m_policy.end();
+    value_type ltmp;
+    init(ltmp);
+    if constexpr (std::is_same<TagType, void>::value) {
+#pragma acc parallel loop gang vector reduction(+ : ltmp) copyin(a_functor)
+      for (auto i = begin; i < end; i++) a_functor(i, ltmp);
+    } else {
+#pragma acc parallel loop gang vector reduction(+ : ltmp) copyin(a_functor)
+      for (auto i = begin; i < end; i++) a_functor(TagType(), i, ltmp);
+    }
+    tmp = ltmp;
+  }
```

Comment:
This could be collapsed into a single function returning the value which would be easier to read. Furthermore, this would reduce the number of times we get `begin`, `end`, etc.

Reply:
Sorry Jakob, I need more details. What do you mean with "this could be collapsed"? All the "reduce" functions in one?

Reply:
I meant, that the body of the `init` function could be put into the reduce function, thus only leaving a single function.

Reply:
Sure! We can do that. Do you want me to do it? There some lines to change here, one per supported reduction type.